scripts/exrrfs_ungrib.sh

```diff
@@ -39,31 +39,33 @@ elif [[ "${prefix}" == "GEFS" ]]; then
   ls -lth ${filea} ${fileb}
   cat ${filea} ${fileb} > GRIBFILE.AAA
 
-elif [[ "${prefix}" == "RAP" ]]; then
-  fstr=$(printf %02d ${FHRin})
-  GRIBFILE=${COMINrap}/rap.${CDATEin:0:8}/rap.t${CDATEin:8:2}z.wrfnatf${fstr}.grib2
-  ${cpreq} ${GRIBFILE} GRIBFILE.AAA
-
-elif [[ "${prefix}" == "RRFS" ]]; then
-  fstr=$(printf %02d ${FHRin})
-  GRIBFILE=${COMINrrfs1}/rrfs_a.${CDATEin:0:8}/${CDATEin:8:2}/rrfs.t${CDATEin:8:2}z.natlve.f${fstr}.grib2
-  ${cpreq} ${GRIBFILE} GRIBFILE.AAA
-
 else
-  echo "ungrib PREFIX=${prefix} not supported"
-  err_exit
+  echo "ungrib PREFIX=${prefix} from xml"
 fi
 #
 # generate the namelist on the fly
-CDATEout=$($NDATE ${FHR} ${CDATE})
-start_time=$(date -d "${CDATEout:0:8} ${CDATEout:8:2}" +%Y-%m-%d_%H:%M:%S) 
-end_time=${start_time}
-interval_seconds=3600
 if [[ "${MESH_NAME}" == "conus3km"   ]]; then
   dx=3; dy=3
 else
   dx=12; dy=12
 fi
+
+fhr_all=$(seq $((10#${OFFSET})) $((10#${INTERVAL})) $(( 10#${OFFSET} + 10#${LENGTH})) )
+for fhr in  ${fhr_all}; do
+
+  HHH=$(printf %03d ${fhr})
+  DATAH=${DATA}/${HHH}
+  mkdir -p ${DATAH}
+  cd ${DATAH}
+  ${cpreq} ${FIXrrfs}/ungrib/Vtable.${prefix} Vtable
+  NAME_FILE=$(echo "${NAME_PATTERN}" | sed "s/\${HHH}/${HHH}/g")
+  GRIBFILE="${COMIN}/${NAME_FILE}"
+  ${cpreq} ${GRIBFILE} GRIBFILE.AAA
+
+CDATEout=$($NDATE ${fhr} ${CDATEin})
+start_time=$(date -d "${CDATEout:0:8} ${CDATEout:8:2}" +%Y-%m-%d_%H:%M:%S) 
+end_time=${start_time}
+interval_seconds=$(( ${INTERVAL}*3600 ))
```

Comment:
If we don't want to use meta tasks to do ungrib for each file, we can run one ungrib.x to generate all needed files instead of launching ungrib.x multiple times.

Reply:
Agree, we can run one ungrid.exe. This need more work on scripts. Should make such change in the future.

Reply:
@hu5970 Here is [my previous scripts](https://raw.githubusercontent.com/RRFSx/rrfs-workflow_beta/c17831ec4165759288b3dc57fd2b63e836d6b98d/scripts/exrrfs_lbc.sh) to run one ungrib.x. The relevant part:

```
nfiles=$((10#$lbc_length/10#$lbc_interval))

for knt in $(seq 1 $((10#$nfiles+1)) ); do
  FHR=$( printf %03d $((10#$lbc_offset+(10#$knt-1)*10#$lbc_interval )) )
  ${cpreq} ${COMINgfs}/gfs.${CDATElbc:0:8}/${CDATElbc:8:2}/gfs.t${CDATElbc:8:2}z.pgrb2.0p25.f${FHR}  $(${USHrrfs}/num_to_GRIBFILE.XXX.sh ${knt})
done
```
You can try `${USHrrfs}/num_to_GRIBFILE.XXX.sh` at the command line, for example:
```
[Jet|fe3 ush]$ num_to_GRIBFILE.XXX.sh 04
GRIBFILE.AAD
```

Reply:
Changed based on guoqing's ${USHrrfs}/num_to_GRIBFILE.XXX.sh .